### DISCUSSION TITLE
Disappearing id boundary line when looking in front

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#273] [Q&A] [answered]
### Malt version

Development-latest

### Blender version

Blender 3.1

### OS

Windows 10 64bits

### Hardware info

Intel i5-9400F | 8 GB RAM | NVDIA GTX 1660 Ti

### Issue description and reproduction steps

Please check the disappearing boundary line when the camera angle is changed from 89 to 90

![image](https://user-images.githubusercontent.com/36077492/158596153-c07dd76f-82a4-4ffc-9620-30ccdfbb164a.png)

![image](https://user-images.githubusercontent.com/36077492/158596304-f6061af1-b79b-4656-b37e-099adfb32ebe.png)

### Attachments

[id boundary line.zip](https://github.com/bnpr/Malt/files/8262344/id.boundary.line.zip)

**Comment by pragma37** [comment, marked as answer]
(I'm converting this into a question, since this is more a current design limitation than a bug.)

The line detection function only takes into account pixels that are further from the camera than the currently shaded pixel.
In the case of a plane viewed at 90 degrees, all pixels have the same depth so all the plane pixels are discarded.

The internal line function has different "depth modes": NEAR, FAR, and ANY, but they're not exposed in the line detection nodes since the Malt UI doesn't support enums yet.

As a workaround, you could add a custom function that works in ANY mode:
```glsl
LineDetectionOutput line_detection_any_depth()
{
    LineDetectionOutput result;

    #ifdef NPR_FILTERS_ACTIVE
    {
        result = line_detection(
            POSITION,
            NORMAL, NORMAL, 
            1,
            LINE_DEPTH_MODE_ANY,
            screen_uv(),
            IN_NORMAL_DEPTH,
            3,
            IN_NORMAL_DEPTH,
            IN_ID
        );
    }
    #endif

    return result;
}
```

The problem with the ANY mode is that lines end up being 1 pixel thicker, since there are 2 nearby lines detected:
![imagen](https://user-images.githubusercontent.com/36610999/158628480-2246bf7c-5f75-41f5-bbbc-17e6c5aeec0d.png)

So you may want to put these "secondary" IDs into the second ID channel, perform the line detection separately and scale it down accordingly.
